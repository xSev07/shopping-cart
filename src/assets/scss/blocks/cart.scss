$totalBGColor             : #f5f5f5;

.cart {
  &__form {
    display: flex;
    flex-wrap: nowrap;
    justify-content: space-between;
    align-items: flex-start;
    max-width: $tabletWidth;
    margin: 0 auto;
  }

  &__list {
    width: 60%;
  }

  &__item {
    display: flex;
    justify-content: space-between;
    margin-bottom: 20px;
  }

  &__picture {
    width: 100px;
    height: 100px;
    margin-right: 25px;
  }

  &__info {
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
  }

  &__title {
    width: 60%;
  }

  &__link {
    color: inherit;

    &:hover,
    &:focus {
      opacity: 0.5;
    }

    &:active {
      opacity: 0.3;
    }

    &:visited {
      color: inherit;
    }
  }

  &__price-wrapper {
    width: 30%;
    text-align: right;
  }

  &__price-total {
    font-weight: bold;
  }

  &__price {
    font-size: smaller;
  }

  &__count-wrapper {
    margin-left: auto;
  }

  &__total {
    width: 30%;
    padding: 30px 10px;
    background: $totalBGColor;
  }

  &__total-row {
    margin-bottom: 15px;
  }

  &__total-count {
    margin-right: 10px;
    font-size: smaller;
  }

  &__total-price {
    flex-shrink: 0;
    font-weight: bold;
  }
}

@media (max-width: $largeTabletWidth) {
  .cart {
    &__form {
      flex-direction: column;
      margin: 0 40px;
    }

    &__list {
      width: 100%;
    }

    &__total {
      padding: 10px;
      display: flex;
      justify-content: space-evenly;
      width: 60%;
      margin: 0 auto;
      background: none;
    }

    &__total-row {
      display: flex;
      flex-direction: column;
      justify-content: space-around;
    }

    &__submit {
      width: 50%;
      margin-left: 10px;
    }
  }
}

@media (max-width: $smTabletWidth) {
  .cart {
    &__total {
      width: 100%;
    }
  }
}


@media (max-width: $phoneWidth) {
  .cart {
    &__form {
      flex-direction: column;
      min-width: 350px;
      margin: 0 20px;
    }
  }
}
