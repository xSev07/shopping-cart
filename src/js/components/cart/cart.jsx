import React from "react";
import PropTypes from "prop-types";
import CartItem from "../cart-item/cart-item.jsx";
import CartAddition from "../cart-addition/cart-addition.jsx";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer";
import {extendObject} from "../../utils/common";
import {allGoodsToSelectedGoods} from "../../adapters/goods";

const Cart = (props) => {
  const {allGoods, selectedGoods, addGoods, deleteGoods} = props;

  return (
    <>
      <CartAddition allGoods={allGoods} onFormSubmit={addGoods}/>
      <section className="cart container">
        <h2 className="visually-hidden">Товары в корзине</h2>
        <form className="cart__form">
          <ul className="cart__list">
            {selectedGoods.map((it) => <CartItem key={it.id} goods={it} onDeleteButtonClick={deleteGoods}/>)}
          </ul>
          <div className="cart__total">
            <div className="cart__row cart__total-row cart-row">
              <p className="cart__total-count">В корзине 2 товара</p>
              <p className="cart__total-price">105 980 &#8381;</p>
            </div>
            <button className="cart__submit cart-button cart-button--attention" type="submit">Продолжить оформление</button>
          </div>
        </form>
      </section>
    </>
  );
};

const mapStateToProps = (state) => ({
  allGoods: state.allGoods,
  selectedGoods: state.selectedGoods,
});

const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const {allGoods, selectedGoods} = stateProps;
  const {dispatch} = dispatchProps;
  const reverseSelectedGoods = selectedGoods.slice().reverse();
  const addGoods = (id, value) => {
    let newSelectedGoods = [];
    const index = selectedGoods.findIndex((it) => it.id === id);
    if (index === -1) {
      const currentGoods = allGoods.find((it) => it.id === id);
      newSelectedGoods = selectedGoods.slice();
      newSelectedGoods.push(allGoodsToSelectedGoods(currentGoods, value));
    } else {
      const currentGoods = selectedGoods[index];
      const newGoods = extendObject(currentGoods, {count: currentGoods.count + value});
      newSelectedGoods = [...selectedGoods.slice(0, index), newGoods, ...selectedGoods.slice(index + 1)];
    }
    dispatch(ActionCreator.changeSelectedGoods(newSelectedGoods));
  };

  const deleteGoods = (id) => {
    const index = selectedGoods.findIndex((it) => it.id === id);
    const newSelectedGoods = [...selectedGoods.slice(0, index), ...selectedGoods.slice(index + 1)];
    dispatch(ActionCreator.changeSelectedGoods(newSelectedGoods));
  };

  return extendObject(
      stateProps,
      dispatchProps,
      ownProps,
      {
        selectedGoods: reverseSelectedGoods,
        addGoods,
        deleteGoods,
      }
  );
};

export {Cart};
export default connect(mapStateToProps, null, mergeProps)(Cart);
