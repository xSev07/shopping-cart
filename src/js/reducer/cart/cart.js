import {extendObject} from "../../utils/common";
import {goods as MockGoods} from "../../mocks/goods";
// import {parseGoodsList} from "./adapters/goods";

const initialState = {
  allGoods: [],
  selectedGoods: [],
};

const ActionType = {
  LOAD_GOODS: `LOAD_GOODS`,
  CHANGE_SELECTED_GOODS: `CHANGE_SELECTED_GOODS`,
};

const ActionCreator = {
  loadGoods: (goods) => ({
    type: ActionType.LOAD_GOODS,
    payload: goods,
  }),
  changeSelectedGoods: (goods) => ({
    type: ActionType.CHANGE_SELECTED_GOODS,
    payload: goods,
  }),
};

const Operation = {
  // eslint-disable-next-line
  loadGoods: () => (dispatch, getState, api) => {
    dispatch(ActionCreator.loadGoods(MockGoods));

    // return api.get(`/goods`)
    //   .then((response) => dispatch(ActionCreator.loadGoods(parseGoodsList(response.data))));
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_GOODS:
      return extendObject(state, {allGoods: action.payload});
    case ActionType.CHANGE_SELECTED_GOODS:
      return extendObject(state, {selectedGoods: action.payload});
  }

  return state;
};

export {reducer, ActionType, ActionCreator, Operation};
