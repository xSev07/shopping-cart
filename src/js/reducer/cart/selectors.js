import NameSpace from "../name-space";
import {createSelector} from "reselect";

const NAME_SPACE = NameSpace.CART;

export const getAllGoods = (state) => {
  return state[NAME_SPACE].allGoods;
};

export const getSelectedGoods = (state) => {
  return state[NAME_SPACE].selectedGoods;
};

export const getGoodsCount = createSelector(
    getSelectedGoods,
    (selectedGoods) => {
      return selectedGoods.reduce((acc, it) => acc + it.count, 0);
    }
);

export const getGoodsTotal = createSelector(
    getSelectedGoods,
    (selectedGoods) => {
      return selectedGoods.reduce((acc, it) => acc + it.price * it.count, 0);
    }
);
