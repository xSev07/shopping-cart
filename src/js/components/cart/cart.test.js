import React from "react";
import renderer from "react-test-renderer";
import {allGoods, selectedGoods} from "../../test-data/test-data";
import {Cart} from "./cart.jsx";

describe(`Snapshot tests for Cart`, () => {
  it(`Should render correctly first load`, () => {
    const tree = renderer
      .create(
          <Cart
            allGoods={allGoods}
            selectedGoods={[]}
            countGoods={0}
            totalGoods={0}
            addGoods={() => {}}
            deleteGoods={() => {}}
            changeGoodsCount={() => {}}
            onFormSubmit={() => {}}
          />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`Should render correctly with selected goods`, () => {
    const tree = renderer
      .create(
          <Cart
            allGoods={allGoods}
            selectedGoods={selectedGoods}
            countGoods={5}
            totalGoods={1350}
            addGoods={() => {}}
            deleteGoods={() => {}}
            changeGoodsCount={() => {}}
            onFormSubmit={() => {}}
          />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
