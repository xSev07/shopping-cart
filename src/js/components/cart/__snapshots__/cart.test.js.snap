// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests for Cart Should render correctly first load 1`] = `
Array [
  <section
    className="cart-addition container"
  >
    <h2
      className="cart-addition__title"
    >
      Добавить товар
    </h2>
    <form
      action="#"
      className="cart-addition__form"
      onSubmit={[Function]}
    >
      <select
        aria-label="Выбрать товар из списка"
        className="cart-addition__select"
      >
        <option
          value="1"
        >
          Видеокарта GIGABYTE GeForce GTX 1050 Ti OC 4G
        </option>
        <option
          value="2"
        >
          Видеокарта GIGABYTE GV-N2060WF2OC-6GD V2
        </option>
        <option
          value="3"
        >
          Видеокарта GIGABYTE GeForce RTX 2080 WINDFORCE OC 8G
        </option>
        <option
          value="4"
        >
          Видеокарта Palit GeForce GT 710 2GB DDR3 Silent
        </option>
        <option
          value="5"
        >
          Видеокарта GIGABYTE GeForce GTX 1080 TI WaterforceWB Xtreme Edition 1
        </option>
      </select>
      <div
        className="cart-count cart-addition__count-wrapper"
      >
        <button
          className="cart-count__button"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
        <input
          aria-label="Количество"
          className="cart-count__value"
          min="1"
          onChange={[Function]}
          type="number"
          value={1}
        />
        <button
          className="cart-count__button"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
      <button
        className="cart-addition__add cart-button cart-button--attention"
        type="submit"
      >
        Добавить
      </button>
    </form>
  </section>,
  <section
    className="cart container"
  >
    <h2
      className="visually-hidden"
    >
      Товары в корзине
    </h2>
    <form
      className="cart__form"
      onSubmit={[Function]}
    >
      <ul
        className="cart__list"
      />
      <div
        className="cart__total"
      >
        <div
          className="cart__row cart__total-row cart-row"
        >
          <p
            className="cart__total-count"
          >
            В корзине 
            0
             
            товаров
          </p>
          <p
            className="cart__total-price"
          >
            0
             ₽
          </p>
        </div>
        <button
          className="cart__submit cart-button cart-button--attention"
          disabled={true}
          type="submit"
        >
          Продолжить оформление
        </button>
      </div>
    </form>
  </section>,
]
`;

exports[`Snapshot tests for Cart Should render correctly with selected goods 1`] = `
Array [
  <section
    className="cart-addition container"
  >
    <h2
      className="cart-addition__title"
    >
      Добавить товар
    </h2>
    <form
      action="#"
      className="cart-addition__form"
      onSubmit={[Function]}
    >
      <select
        aria-label="Выбрать товар из списка"
        className="cart-addition__select"
      >
        <option
          value="1"
        >
          Видеокарта GIGABYTE GeForce GTX 1050 Ti OC 4G
        </option>
        <option
          value="2"
        >
          Видеокарта GIGABYTE GV-N2060WF2OC-6GD V2
        </option>
        <option
          value="3"
        >
          Видеокарта GIGABYTE GeForce RTX 2080 WINDFORCE OC 8G
        </option>
        <option
          value="4"
        >
          Видеокарта Palit GeForce GT 710 2GB DDR3 Silent
        </option>
        <option
          value="5"
        >
          Видеокарта GIGABYTE GeForce GTX 1080 TI WaterforceWB Xtreme Edition 1
        </option>
      </select>
      <div
        className="cart-count cart-addition__count-wrapper"
      >
        <button
          className="cart-count__button"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
        <input
          aria-label="Количество"
          className="cart-count__value"
          min="1"
          onChange={[Function]}
          type="number"
          value={1}
        />
        <button
          className="cart-count__button"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
      <button
        className="cart-addition__add cart-button cart-button--attention"
        type="submit"
      >
        Добавить
      </button>
    </form>
  </section>,
  <section
    className="cart container"
  >
    <h2
      className="visually-hidden"
    >
      Товары в корзине
    </h2>
    <form
      className="cart__form"
      onSubmit={[Function]}
    >
      <ul
        className="cart__list"
      >
        <li
          className="cart__item"
        >
          <img
            alt="Видеокарта GIGABYTE GeForce GTX 1050 Ti OC 4G"
            className="cart__picture"
            height="100"
            src="https://img.mvideo.ru/Pdb/30031080m.jpg"
            width="100"
          />
          <div
            className="cart__info"
          >
            <div
              className="cart__row cart-row"
            >
              <h3
                className="cart__title"
              >
                <a
                  className="cart__link"
                  href="#"
                >
                  Видеокарта GIGABYTE GeForce GTX 1050 Ti OC 4G
                </a>
              </h3>
              <div
                className="cart__price-wrapper"
              >
                <p
                  className="cart__price-total"
                >
                  25,380
                   ₽
                </p>
                <p
                  className="cart__price"
                >
                  2
                   x 
                  12,690
                   ₽
                </p>
              </div>
            </div>
            <div
              className="cart__row cart-row"
            >
              <button
                className="cart__delete cart-button  cart-button--in-text"
                onClick={[Function]}
                type="button"
              >
                Удалить
              </button>
              <div
                className="cart-count cart__count-wrapper cart-row"
              >
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  -
                </button>
                <input
                  aria-label="Количество"
                  className="cart-count__value"
                  min="1"
                  onChange={[Function]}
                  type="number"
                  value={2}
                />
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </li>
        <li
          className="cart__item"
        >
          <img
            alt="Видеокарта GIGABYTE GV-N2060WF2OC-6GD V2"
            className="cart__picture"
            height="100"
            src="https://img.mvideo.ru/Pdb/30052540m.jpg"
            width="100"
          />
          <div
            className="cart__info"
          >
            <div
              className="cart__row cart-row"
            >
              <h3
                className="cart__title"
              >
                <a
                  className="cart__link"
                  href="#"
                >
                  Видеокарта GIGABYTE GV-N2060WF2OC-6GD V2
                </a>
              </h3>
              <div
                className="cart__price-wrapper"
              >
                <p
                  className="cart__price-total"
                >
                  29,490
                   ₽
                </p>
              </div>
            </div>
            <div
              className="cart__row cart-row"
            >
              <button
                className="cart__delete cart-button  cart-button--in-text"
                onClick={[Function]}
                type="button"
              >
                Удалить
              </button>
              <div
                className="cart-count cart__count-wrapper cart-row"
              >
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  -
                </button>
                <input
                  aria-label="Количество"
                  className="cart-count__value"
                  min="1"
                  onChange={[Function]}
                  type="number"
                  value={1}
                />
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </li>
        <li
          className="cart__item"
        >
          <img
            alt="Видеокарта GIGABYTE GeForce RTX 2080 WINDFORCE OC 8G"
            className="cart__picture"
            height="100"
            src="https://img.mvideo.ru/Pdb/30040867m.jpg"
            width="100"
          />
          <div
            className="cart__info"
          >
            <div
              className="cart__row cart-row"
            >
              <h3
                className="cart__title"
              >
                <a
                  className="cart__link"
                  href="#"
                >
                  Видеокарта GIGABYTE GeForce RTX 2080 WINDFORCE OC 8G
                </a>
              </h3>
              <div
                className="cart__price-wrapper"
              >
                <p
                  className="cart__price-total"
                >
                  150,000
                   ₽
                </p>
                <p
                  className="cart__price"
                >
                  3
                   x 
                  50,000
                   ₽
                </p>
              </div>
            </div>
            <div
              className="cart__row cart-row"
            >
              <button
                className="cart__delete cart-button  cart-button--in-text"
                onClick={[Function]}
                type="button"
              >
                Удалить
              </button>
              <div
                className="cart-count cart__count-wrapper cart-row"
              >
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  -
                </button>
                <input
                  aria-label="Количество"
                  className="cart-count__value"
                  min="1"
                  onChange={[Function]}
                  type="number"
                  value={3}
                />
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </li>
        <li
          className="cart__item"
        >
          <img
            alt="Видеокарта Palit GeForce GT 710 2GB DDR3 Silent"
            className="cart__picture"
            height="100"
            src="https://img.mvideo.ru/Pdb/30030965m.jpg"
            width="100"
          />
          <div
            className="cart__info"
          >
            <div
              className="cart__row cart-row"
            >
              <h3
                className="cart__title"
              >
                <a
                  className="cart__link"
                  href="#"
                >
                  Видеокарта Palit GeForce GT 710 2GB DDR3 Silent
                </a>
              </h3>
              <div
                className="cart__price-wrapper"
              >
                <p
                  className="cart__price-total"
                >
                  10,000
                   ₽
                </p>
              </div>
            </div>
            <div
              className="cart__row cart-row"
            >
              <button
                className="cart__delete cart-button  cart-button--in-text"
                onClick={[Function]}
                type="button"
              >
                Удалить
              </button>
              <div
                className="cart-count cart__count-wrapper cart-row"
              >
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  -
                </button>
                <input
                  aria-label="Количество"
                  className="cart-count__value"
                  min="1"
                  onChange={[Function]}
                  type="number"
                  value={1}
                />
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </li>
        <li
          className="cart__item"
        >
          <img
            alt="Видеокарта GIGABYTE GeForce GTX 1080 TI WaterforceWB Xtreme Edition 1"
            className="cart__picture"
            height="100"
            src="https://img.mvideo.ru/Pdb/30031081m.jpg"
            width="100"
          />
          <div
            className="cart__info"
          >
            <div
              className="cart__row cart-row"
            >
              <h3
                className="cart__title"
              >
                <a
                  className="cart__link"
                  href="#"
                >
                  Видеокарта GIGABYTE GeForce GTX 1080 TI WaterforceWB Xtreme Edition 1
                </a>
              </h3>
              <div
                className="cart__price-wrapper"
              >
                <p
                  className="cart__price-total"
                >
                  25,000
                   ₽
                </p>
              </div>
            </div>
            <div
              className="cart__row cart-row"
            >
              <button
                className="cart__delete cart-button  cart-button--in-text"
                onClick={[Function]}
                type="button"
              >
                Удалить
              </button>
              <div
                className="cart-count cart__count-wrapper cart-row"
              >
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  -
                </button>
                <input
                  aria-label="Количество"
                  className="cart-count__value"
                  min="1"
                  onChange={[Function]}
                  type="number"
                  value={1}
                />
                <button
                  className="cart-count__button"
                  onClick={[Function]}
                  type="button"
                >
                  +
                </button>
              </div>
            </div>
          </div>
        </li>
      </ul>
      <div
        className="cart__total"
      >
        <div
          className="cart__row cart__total-row cart-row"
        >
          <p
            className="cart__total-count"
          >
            В корзине 
            5
             
            товаров
          </p>
          <p
            className="cart__total-price"
          >
            1,350
             ₽
          </p>
        </div>
        <button
          className="cart__submit cart-button cart-button--attention"
          disabled={false}
          type="submit"
        >
          Продолжить оформление
        </button>
      </div>
    </form>
  </section>,
]
`;
